app/src/main/java/com/keylesspalace/tusky/components/compose/dialog/FocusDialog.kt

```diff
@@ -0,0 +1,235 @@
+/* Copyright 2019 Tusky Contributors
+ *
+ * This file is a part of Tusky.
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>. */
+
+package com.keylesspalace.tusky.components.compose.dialog
+
+import android.app.Activity
+import android.content.DialogInterface
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Matrix
+import android.graphics.Paint
+import android.graphics.Path
+import android.graphics.drawable.Drawable
+import android.net.Uri
+import android.os.Build
+import android.view.WindowManager
+import android.widget.ImageView
+import android.widget.LinearLayout
+import android.widget.Toast
+import androidx.appcompat.app.AlertDialog
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.lifecycleScope
+import at.connyduck.sparkbutton.helpers.Utils
+import com.bumptech.glide.Glide
+import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
+import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
+import com.bumptech.glide.load.resource.bitmap.DownsampleStrategy
+import com.bumptech.glide.load.resource.bitmap.TransformationUtils
+import com.bumptech.glide.request.target.CustomTarget
+import com.bumptech.glide.request.transition.Transition
+import com.bumptech.glide.util.Util
+import com.github.chrisbanes.photoview.OnPhotoTapListener
+import com.github.chrisbanes.photoview.PhotoView
+import com.keylesspalace.tusky.R
+import com.keylesspalace.tusky.entity.Attachment.Focus
+import kotlinx.coroutines.launch
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+import java.util.concurrent.locks.Lock
+
+// Private, but necessary to implement BitmapTransformation, function extracted from Glide
```

Comment:
This file now contains rather a lot of code adapted from Glide. There was no sample code for creating a custom BitmapTransformation in their docs so I used the RoundedCorners class and the TransformationUtils.roundedCorners method as bases. `getAlphaSafeBitmap` and `getAlphaSafeConfig` are taken direct from Glide TransformationUtils (but converted to Kotlin) and the "boilerplate" methods are almost identical to the ones from RoundedCorners. I do not think this introduces license issues because (1) we are already bound to the Glide license and (2) the code taken is "minimal" in the sense that it practically could not be any different than it is.

I am going to file a bug on Glide asking them to make getAlphaSafeBitmap and getAlphaSafeConfig public methods as I do not think it is possible to implement a BitmapTransformation without them.